README.md

```diff
@@ -1,16 +1,644 @@
+│ │                         │      │          │
+│ └─────────────────────┬───┘      │          │
+│                       │          │          │
+│                       │          │          │
+│              ┌────────▼──────────▼────────┐ │
+│              │                            │ │
+│              │ https://example.com/events │ │
+│              │         crud/create        │ │
+│              │                            │ │
+│              └────────────────────────────┘ │
+│                                             │
+└─────────────────────────────────────────────┘
+```
+
+ ``` json
+{
+  "ucan/invoke": "Qmd4trNUhgWwsBbSsYBEWJqgiHyLrnhZ8u1DJsWsEKeuuF",
+  "v": "0.1.0",
+  "nnc": "abcdef",
+  "ext": null,
+  "run": {
+    "update-dns": {
+      "using": "dns://example.com?TYPE=TXT",
+      "do": "crud/update"
+      "inputs": [
+        { 
+          "value": "hello world",
+          "retries": 5
+        }
+      ]
+    }
+  },
+  "sig": "kQHtTruysx4S8SrvSjTwr6ttTLzc7dd7atANUYT-SRK6v_SX8bjHegWoDak2x6vTAZ6CcVKBt6JGpgnjABpsoL"
+}
+ 
+{
+  "ucan/invoke": "QmbXdT8QQMJ55Lb6MGJqTmwNzuUnHsE18t7zXGWeq9rQcV",
+  "v": "0.1.0",
+  "nnc": "12345",
+  "ext": null,
+  "run": {
+    "notify-carol": {
+      "using": "mailto://alice@example.com",
+      "do": "msg/send",
+      "inputs": [
+        {
+          "to": "carol@example.com",
+          "subject": "DNSLink for example.com",
+          "body": {"ucan/promise": ["QmcKi4Z1L7VH4HahyrXb8Rzjex5U7DiY12PnTnysxviuDT", "update-dns", "$http.body"]}
+        }
+      ]
+    }
+  },
+  "sig": "XZRSmp5cHaXX6xWzSTxQqC95kQHtTruysx4S8SrvSjTwr6ttTLzc7dd7atANUQJXoWThUiVuCHWdMnQNQJgiJi"
+}
+
+{
+  "ucan/invoke": "QmY4jEVE35u8SHegWoDak2x6vTAZ6Cc4cpSD5LAUQ23W7L",
+  "v": "0.1.0",
+  "nnc": "02468",
+  "ext": null,
+  "run": {
+    "notify-bob": {
+      "using": "mailto://alice@example.com",
+      "do": "msg/send",
+      "inputs": [
+        {
+          "to": "bob@example.com",
+          "subject": "DNSLink for example.com",
+          "body": {"ucan/promise": ["QmcKi4Z1L7VH4HahyrXb8Rzjex5U7DiY12PnTnysxviuDT", "update-dns", "$http.body"]}
+        }
+      ]
+    },
+    "log-as-done": {
+      "using": "https://example.com/report"
+      "do": "crud/update"
+      "inputs": [
+        {
+          "from": "mailto://alice@exmaple.com",
+          "to": ["bob@exmaple.com", "carol@example.com"],
+          "event": "email-notification",
+        },
+        {
+          "_": {"ucan/promise": ["/", "notify-bob"]}
+        },
+        {
+          "_": {"ucan/promise": ["QmaCRSugy2sXEKxjDDwcqLx2Bs7CUGpZSrtsASsNyntyC9", "notify-carol"]}
+        }
+      ]
+    }
+  },
+  "sig": "5vNn4--uTeGk_vayyPuNTYJ71Yr2nWkc6AkTv1QPWSgetpsu8SHegWoDakPVTdxkWb6nhVKAz6JdpgnjABppC7"
+}
+```
+
+# 6 Appendix
+
+## 6.1 Support Types
+
+``` ipldsch
+type CID = String
+type URI = String
+type Ability = String
+type Path = String
+
+type SemVer struct {
+  num NumVer
+  tag String (implicit Null)
```

Comment:
I think you may want optional as opposed to nullable here

Reply:
I find that section of the IPLD Schema docs pretty confusing! My understanding was that `optional` can be combined with `nullable` for parsing, but that `implicit Null` would always fill this field in on serialization, which makes omitting the field the same as setting it to Null. But as I say, its' not clear tome yet.

Is the above interpretation correct? If not, how does it work? 🙏

Reply:
> I find that section of the IPLD Schema docs pretty confusing! My understanding was that optional can be combined with nullable for parsing, but that implicit Null would always fill this field in on serialization, which makes omitting the field the same as setting it to Null. But as I say, its' not clear tome yet.

I believe that to be the case as well.

I have said you may want to use optional instead because  of the `stringjoin` directive below, if it's implicitly Null I think it may append `+Null` or something, but if it's optional and omitted I would expect it not add `+` at all.

That said I actually have no idea what it is supposed to do if it is either `Null` or omitted. Perhaps @rvagg can help us understand this better ?

Reply:
This specific one being `optional` does make sense, yeah. I still would love it if @rvagg could make the distinction clearer, though!

Reply:
* `optional` means that the entry can be omitted entirely from the encoded form
* `nullable` means that the field could have the value `Null` as well as the type specified
* combining both means you get the extra cardinality of `Null` plus the possibility of it being omitted from the encoded form, it's kind of a :shrug: this may or may not be there and it may or may not be null, not exactly awesome unless you're dealing with existing data that is loose like that
* the point about `implicit Null` is interesting and I can't think of a case where we're using that; perhaps the only place where it's useful is where you have `nullable optional`?
  * What it would be saying is "when encoding this, if it's missing, make it `Null`, but when decoding this, accept a value, `Null` or for it to be missing.
  * What it does coming out of a decode is interesting though and I'm not sure we have defined behaviour for this—we have this notion of "Absent" for fields that are not set, which is distinct from a `Null` value. In JS an Absent is emulated as an undefined (and unset) field in your object vs a `null`.
  * `(implicit Null)` won't be accepted by either the JS or Go DSL parsers, perhaps it should be? Currently we only accept Strings, Ints, Floats (maybe) or Bools in this position, but we also need to add Bytes at some stage. Null is an interesting case we may need to consider?
  * 🤔 
* `optional` becomes tricky when you're not encoding a struct as a map, so avoid it when doing `tuple`, or `stringjoin` or something like that; we have said in the past that it should be possible to have the last element of your struct be `optional` when doing a tuple encoding but I'm not sure we have any implementations (or maybe even docs) that do this.

Not sure any of that helps; but let me know if you're keen to take the `implicit Null` thing further; that might be something we should have some test fixtures for if we want to allow it.

Reply:
> What it does coming out of a decode is interesting though and I'm not sure we have defined behaviour for this—we have this notion of "Absent" for fields that are not set, which is distinct from a Null value. In JS an Absent is emulated as an undefined (and unset) field in your object vs a null.

It's entirely possible that I have too much Haskell brain at the type level here. Without the `Null` on an optional field, you essentially have `undefined`, which is _yet another_ state. What `undefined` "means" is often really unclear (famously: "The Billion Dollar Mistake"). When designing types for untagged optionals, I basically always define it as `null`, which does have a definite meaning (it's a value, not the absence of one).

> (implicit Null) won't be accepted by either the JS or Go DSL parsers,

Oh wow, that's really good to know! So I can't `implicit {}` or `implicit []` either I guess?

> Not sure any of that helps; but let me know if you're keen to take the implicit Null thing further; 

Disclaimer: I recognize that I come at this stuff from an atypical, very strongly typed ("function programming weirdo") background. _I_ found the above information surprising.

For my use cases, I would like to have the following...

```
implicit Null
implicit {}
implicit []
```

...and am happy to help make that happen! 👍 What would the next steps be if you're open to that?

Also: thanks a bunch for the quick and thorough replies @rvagg! 🙏 

# IPLD Sets

Off topic, but while we're adding stuff to IPLD 😉 Something that @Gozala mentioned earlier is that also having a `Set` (sorted array) type would be useful. We keep describing it in prose, but don't have a way to describe that in IPLD.

Reply:
I think Eric would agree with you on the `undefined` thing, and I'm generally inclined toward explicitness so `optional` isn't very comfy. The `Absent` thing helps disambiguate the `optional` case for typed structs in Go where that field exists but shouldn't be `Null`. IMO `optional` is mainly helpful in adapting to legacy forms or where you care a lot about saving bytes.

To combine your other notes on implicits and Sets, see my comment above where you discussed Sets that got lost in the Resolved issue: https://github.com/ucan-wg/invocation/pull/1#discussion_r1035109958, repeated here for visibility:

> We really need sets in IPLD data format

('scuse another snipe, but this is an interesting topic)

```
type Void unit representation emptymap
type Set {String:Void}`
```

would be a way of expressing and encoding a set... a map where the keys are what matters .. of course we're a bit limited by the fact that map keys need to be strings _but_ if something has a representation that encodes as a string (`stringjoin`, `stringpairs`, `stringprefix`) then that's valid as a map key. In theory if your map is an ADL then the String keys thing doesn't need to hold true, it's just not something Schemas properly allow you to do but HAMTs make really nice Sets with unit values; as long as you can express the keys as bytes to hash.

The `Unit` type can be `null`, `false`, `true`, `{}`, see https://github.com/ipld/ipld/blob/master/specs/schemas/schema-schema.ipldsch#L753-L789 - one of the newer features of Schemas and I don't think we have any implementation that does anything with it yet but we'll eventually need to work on that cause it's so useful.

---

So .. the `implicit` values you want really are units; which are probably comfortable to a functional programmer. A case could be made that they could occupy an `implicit` where a field is `nullable`; but it might get a bit messy if the type for that field could also be that form .. so we'd need lots of rules around it.

Reply:
re taking things further:

* making changes to how IPLD Schemas work would mean editing the schema-schema; see how `implicit` is defined here: https://github.com/ipld/ipld/blob/373b8ebad58b7ef49ae389ad98f60da1a9bc359f/specs/schemas/schema-schema.ipldsch#L646 (Null doesn't really fit there; in fact Null occupies a weird place in schemas where it's not really a TypeKind, which makes a bunch of things difficult - I ran into this when making a [TypeScript version of schema-schema](https://github.com/ipld/js-ipld-schema/blob/master/schema-schema.ts) and have been thinking of trying to add it)
* making changes to how DSL is parsed depends on where it matters to you - https://github.com/ipld/js-ipld-schema/blob/master/ipld-schema.pegjs & https://github.com/ipld/go-ipld-prime/blob/master/schema/dsl/parse.go are the two main parsers and they are driven by the schema-schema, so if it's in the schema-schema then the parsers _should_ deal with it (but they are both not quite there)
* making changes beyond that is a matter of how you want to consume it! there's a JS `typed` implementation that could be adjusted to fit new changes in (in js-ipld-schema), then there's Go codegen (which has been deprioritised lately) and Go bindnode (which we mostly focus on with Go).